lms/extractors/ziparchive.py

```diff
@@ -30,35 +30,60 @@ def can_extract(self) -> bool:
         return self.is_zipfile
 
     @staticmethod
-    def _extract(archive: ZipFile, filename: str) -> File:
+    def _extract(archive: ZipFile, filename: str, dirname: str = '') -> File:
         with archive.open(filename) as current_file:
             log.debug(f'Extracting from archive: {filename}')
             code = current_file.read()
         decoded = code.decode('utf-8', errors='replace').replace('\x00', '')
-        return File(path=f'/{filename}', code=decoded)
-
-    def get_exercise(self, file: ZipFile) -> Tuple[int, List[File]]:
+        return File(path=f'/{filename.lstrip(dirname)}', code=decoded)
+
+    def get_files(
+        self, archive: ZipFile, namelist: List[Text], dirname: str = '',
+    ) -> List[File]:
+        unwanted_files = self.get_unwanted_files(namelist)
+        return [
+            self._extract(archive, filename, dirname)
+            for filename in namelist
+            if (
+                filename.startswith(dirname)
+                and filename not in unwanted_files
+                and filename != dirname
+            )
+        ]
+
+    def get_exercises_by_dirs(
+        self, archive: ZipFile, namelist: List[Text],
+    ) -> Iterator[Tuple[int, List[File]]]:
+        for dirname in namelist:
+            exercise_id, _ = self._clean(dirname.rpartition(os.path.sep)[0])
```

Comment:
Can you please use `parent_name, filename = os.path.split(dirname)` instead?

It's more concise and will probably prevent us dealing with some strange FS path edge cases.